### PR TITLE
Make a trailing colon for stanzas a parse failure

### PR DESCRIPTION
**Pull request by 9999years**
Here's a mistake I make semi-regularly:

    source-repository-package:
        type: git
        location: https://github.com/parsonsmatt/foundation
        tag: 688c32ccd9a951bc96dd09423a6e6684f091d510
        subdir: basement
        subdir: foundation

Cabal treats this as a warning, so it prints:

    Warning: cabal.project: Unrecognized field
    'source-repository-package' on line 52

This is fine (if you already know the mistake you've made, at least!), but it's very easy to miss amidst lots of output.

I often re-run `cabal` when I see a ton of output to attempt to get a smaller error message. (Usually it works and I get an error message that's got less "compiling module such and such" noise in it.) However, re-running `cabal` will discard this warning entirely!

Let's make it a hard error instead. This is a backwards-compatibility break.

Now, it prints this error:

```
Error: [Cabal-7090]
Error parsing project file cabal.project:52:
'source-repository-package' is a stanza, not a field. Remove the trailing ':' to parse a stanza.
```

- [x] Still needs tests!

### CODE REVIEW

File: cabal-testsuite/PackageTests/ProjectConfig/FieldStanzaConfusion/cabal.test.hs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import Test.Cabal.Prelude
+
+main = cabalTest $ do
+  result <- fails $ cabal' "build" []
+  assertOutputContains "Error parsing project file" result
```
Comment (geekosaur) [inline cabal-testsuite/PackageTests/ProjectConfig/FieldStanzaConfusion/cabal.test.hs:5]: I wonder if these should start using error numbers. (That might be a separate PR, though.)

**Review by geekosaur — APPROVED** [review]

File: cabal-testsuite/PackageTests/ProjectConfig/FieldStanzaConfusion/cabal.test.hs (hunk 1 above)
Comment (9999years) [reply cabal-testsuite/PackageTests/ProjectConfig/FieldStanzaConfusion/cabal.test.hs:5]: It's error 7090 (`ReportParseResult`). Not super specific but there is a number for it.

Comment (geekosaur) [reply cabal-testsuite/PackageTests/ProjectConfig/FieldStanzaConfusion/cabal.test.hs:5]: Oh. Perhaps future work could be making parse errors more specific; one error number for all of them seems like it would make `errors.haskell.org` less useful.

**Comment by Mikolaj** [comment]
I always forget: `cabal.project` format is not guarded under the `cabal-version`, only the .cabal file format is, right?

**Comment by philderbeast** [comment]
Yes @Mikolaj, there's no cabal versioning in project files, neither a version stamp nor a conditional check for cabal's version, #8933 and #10386.

**Comment by Mikolaj** [comment]
@philderbeast: thank you and these are indeed important tickets. So, this breaking change seems fine to be included in the next major version. I don't think it's breaking a useful enough behaviour that an advance warning or deprecation period is needed.

**Review by Mikolaj — APPROVED** [review]
LGTM

**Comment by geekosaur** [comment]
Spot-check… Mergify's still merging while CI is still running. ☹
